vpr/src/pack/post_routing_pb_pin_fixup.cpp

```diff
@@ -0,0 +1,672 @@
+
+        VTR_LOGV(verbose,
+                 "Remap clustered block '%s' global net '%s' to pin '%s'\n",
+                 clustering_ctx.clb_nlist.block_pb(blk_id)->name,
+                 atom_ctx.nlist.net_name(global_atom_net_id).c_str(),
+                 unused_pb_graph_pin->to_string().c_str());
+    }
+
+    /* Replace old pb_routes with the new one */
+    pb->pb_route = new_pb_routes;
+}
+
+/********************************************************************
+ * Top-level function to synchronize a packed netlist to routing results
+ * The problem comes from a mismatch between the packing and routing results
+ * When there are equivalent input/output for any grids, router will try
+ * to swap the net mapping among these pins so as to achieve best 
+ * routing optimization.
+ * However, it will cause the packing results out-of-date as the net mapping 
+ * of each grid remain untouched once packing is done.
+ * This function aims to fix the mess after routing so that the net mapping
+ * can be synchronized
+ *
+ * Note:
+ *   - This function SHOULD be run ONLY when routing is finished!!!
+ *******************************************************************/
+void sync_netlists_to_routing(const DeviceContext& device_ctx,
+                              AtomContext& atom_ctx,
+                              ClusteringContext& clustering_ctx,
+                              const PlacementContext& placement_ctx,
+                              const RoutingContext& routing_ctx,
+                              const bool& verbose) {
+    vtr::ScopedStartFinishTimer timer("Synchronize the packed netlist to routing optimization");
+
+    /* Reset the database for post-routing clb net mapping */
+    clustering_ctx.post_routing_clb_pin_nets.clear();
+    clustering_ctx.pre_routing_net_pin_mapping.clear();
+
+    /* Create net-to-rr_node mapping */
+    vtr::vector<RRNodeId, ClusterNetId> rr_node_nets = annotate_rr_node_nets(device_ctx,
+                                                                             clustering_ctx,
+                                                                             routing_ctx,
+                                                                             verbose);
+
+    /* Update the core logic (center blocks of the FPGA) */
+    for (size_t x = 1; x < device_ctx.grid.width() - 1; ++x) {
+        for (size_t y = 1; y < device_ctx.grid.height() - 1; ++y) {
+            /* Bypass the EMPTY tiles */
+            if (true == is_empty_type(device_ctx.grid[x][y].type)) {
+                continue;
+            }
+            /* Only update multi-width multi-height block once! */
+            if ((0 != device_ctx.grid[x][y].width_offset)
+                || (0 != device_ctx.grid[x][y].height_offset)) {
+                continue;
+            }
+            /* Get the mapped blocks to this grid */
+            for (const ClusterBlockId& cluster_blk_id : placement_ctx.grid_blocks[x][y].blocks) {
```

Comment:
Why do a x/y/blocks loops versus just using `clb_netlist.blocks()` (https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/master/vpr/src/base/netlist.h#L697)?

Reply:
That is a good idea. I will adapt that.